keps/sig-scheduling/4815-dra-partitionable-devices/README.md

```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
-
-	// Attributes defines the set of attributes for this device.
-	// The name of each attribute must be unique in that set.
-	//
-	// To ensure this uniqueness, attributes defined by the vendor
-	// must be listed without the driver name as domain prefix in
-	// their name. All others must be listed with their domain prefix.
-	//
-	// Conflicting attributes from those provided via mixins are
-	// overwritten by the ones provided here.
-	//
-	// The maximum number of attributes and capacities combined is 32.
-	//
-	// +optional
-	Attributes map[QualifiedName]DeviceAttribute `json:"attributes,omitempty"`
-
-	// Capacity defines the set of capacities for this device.
-	// The name of each capacity must be unique in that set.
-	//
-	// To ensure this uniqueness, capacities defined by the vendor
-	// must be listed without the driver name as domain prefix in
-	// their name. All others must be listed with their domain prefix.
-	//
-	// Conflicting capacities from those provided via mixins are
-	// overwritten by the ones provided here.
-	//
-	// The maximum number of attributes and capacities combined is 32.
-	//
-	// +optional
-	Capacity map[QualifiedName]DeviceCapacity `json:"capacity,omitempty"`
+  // Includes defines the set of device mixins that this device includes.
+  //
+  // The propertes of each included mixin are applied to this device in
+  // order. Conflicting properties from multiple mixins are taken from the
+  // last mixin listed that contains them. Properties set on the device will
+  // always override properties from mixins.
+  //
+  // The mixins referenced here must be defined in the same
+  // ResourceSlice.
+  //
+  // The maximum number of mixins that can be included is 8.
+  //
+  // +optional
+  // +listType=atomic
+  Includes []DeviceMixinRef
+
+  // Attributes defines the set of attributes for this device.
+  // The name of each attribute must be unique in that set.
+  //
+  // To ensure this uniqueness, attributes defined by the vendor
+  // must be listed without the driver name as domain prefix in
+  // their name. All others must be listed with their domain prefix.
+  //
+  // Conflicting attributes from those provided via mixins are
+  // overwritten by the ones provided here.
+  //
+  // The maximum number of attributes and capacities combined is 32.
+  //
+  // +optional
+  Attributes map[QualifiedName]DeviceAttribute
+
+  // Capacity defines the set of capacities for this mixin.
+  // The name of each capacity must be unique in that set.
+  //
+  // To ensure this uniqueness, capacities defined by the vendor
+  // must be listed without the driver name as domain prefix in
+  // their name. All others must be listed with their domain prefix.
+  //
+  // Conflicting capacities from those provided via other mixins are
+  // overwritten by the ones provided here.
+  //
+  // The maximum number of attributes and capacities combined is 32.
+  //
+  // +optional
+  Capacity map[QualifiedName]DeviceCapacity
```

Comment:
As I mentiond in another comment. I don't think devices should have the ability to define their own intrinsic capacities. *All* capacity of a device should be pulled from (one or more) pools.

In my mental model with this new pool design, there would be AttributeMixins and CapacityMixins. Devices can  include AttributeMixins and CapacityMixins, but ResourcePools can *only* include CapacityMixins.

When constructing a device you then include its set of desired attribute mixins, and then (in a separate field) list the set of CapacityMixins you want attributed to the device -- each of which *must* including a reference to a pool where the capacities listed in that mixin come from.

Reply:
I think that the internal capacity consumed by partitions being the same as the values advertised to consumers is just a coincidence of the MIG case. I don't think that will be true in the general case and I don't want to lock us into it. For example, I can imagine a device that needs some overhead associated with the internal allocation before making the resource available to consumers. Or, even in the MIG case, telling the external consumers about "memory slices" seems odd.

Maybe there's a way to build the API to simplify the MIG case but still allow the general case.

Reply:
Or, someone may build a device that advertises logical resources but consumes physical ones: I can service 1000 requests per second if I consume 1 core, or 2000 requests per second if I use 2 cores. So, internally they are consuming "cores" externally they are advertising "requests per second".

Reply:
So with the discussion offline, I think we are converging on something like:

* DeviceMixin will include both attributes and capacity. Capacity means the advertised resource availability of a device. These can only be referenced from devices and they are referenced by providing only the name.
* CapacityPoolMixin will include only capacity. These can only be referenced from CapacityPools
* SharedCapacityConsumedMixin (we should find a better name) includes the shared capacity consumed, but does not specify a capacity pool. They can only be referenced from devices. Both the name and the capacity pool will need to be provided when they are included.

I will try to get the PR updated tonight.

Reply:
PR has been updated.